Zotero/Scenes/Detail/PDF/ViewModels/PDFReaderActionHandler.swift

```diff
@@ -1396,32 +1374,170 @@ final class PDFReaderActionHandler: ViewModelActionHandler, BackgroundDbProcessi
+                // `AnnotationStateManager` doesn't apply the `blendMode` to created annotations, so it needs to be applied to newly created annotations here.
+                let (_, _, blendMode) = AnnotationColorGenerator.color(from: activeColor, isHighlight: (annotation is PSPDFKit.HighlightAnnotation), userInterfaceStyle: state.interfaceStyle)
+                annotation.blendMode = blendMode ?? .normal
+
+                // Either annotation is new (key not assigned) or the user used undo/redo and we check whether the annotation exists in DB
+                guard annotation.key == nil || state.annotation(for: .init(key: annotation.key!, type: .database)) == nil else {
+                    keptAsIs.append(annotation)
+                    continue
+                }
+
+                let splitAnnotations = splitIfNeeded(annotation: annotation, user: state.displayName)
+
+                guard splitAnnotations.count > 1 else {
+                    keptAsIs.append(annotation)
+                    continue
+                }
+                DDLogInfo("PDFReaderActionHandler: did split annotations into \(splitAnnotations.count)")
+                toRemove.append(annotation)
+                toAdd.append(contentsOf: splitAnnotations)
+            }
+
+            return (keptAsIs, toRemove, toAdd)
+
+            /// Splits annotation if it exceedes position limit. If it is within limit, it returs original annotation.
+            /// - parameter annotation: Annotation to split
+            /// - parameter user: User which created the annotation if it's new
+            /// - returns: Array with original annotation if limit was not exceeded. Otherwise array of new split annotations.
+            func splitIfNeeded(annotation: PSPDFKit.Annotation, user: String) -> [PSPDFKit.Annotation] {
+                if let annotation = annotation as? HighlightAnnotation, let rects = annotation.rects, let splitRects = AnnotationSplitter.splitRectsIfNeeded(rects: rects) {
+                    return createAnnotations(from: splitRects, original: annotation)
+                }
+
+                if let annotation = annotation as? InkAnnotation, let paths = annotation.lines, let splitPaths = AnnotationSplitter.splitPathsIfNeeded(paths: paths) {
+                    return createAnnotations(from: splitPaths, original: annotation)
+                }
+
+                if annotation.key == nil {
+                    annotation.user = user
+                    annotation.customData = [AnnotationsConfig.keyKey: KeyGenerator.newKey]
+                }
+
+                return [annotation]
+
+                func createAnnotations(from splitRects: [[CGRect]], original: HighlightAnnotation) -> [HighlightAnnotation] {
+                    guard splitRects.count > 1 else { return [original] }
+                    return splitRects.map { rects -> HighlightAnnotation in
+                        let new = HighlightAnnotation()
+                        new.rects = rects
```

Comment:
I see we're assigning `user` above but not here or below.

Reply:
Check https://github.com/zotero/zotero-ios/pull/934/commits/5d5c17041c254a25430915a8b15114a413eedd4c